File: Recursion/best_path.c
/******************************************************************************
 *
 *  Find best path in an MxN matrix
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

#define MAX(l, r)       ((l) > (r) ? (l) : (r))

int bestPathRecursion(int *array, int m, int n, int i, int j) {
    if (i == m - 1 && j == n -1) 
        return array[i * n + j];

    if (i == m - 1)
        return array[i * n + j] + bestPathRecursion(array, m, n, i, j + 1);

    if (j == n - 1)
        return array[i * n + j] + bestPathRecursion(array, m, n, i + 1, j);
    
    int down = array[i * n + j] + bestPathRecursion(array, m, n, i + 1, j);
    int right = array[i * n + j] + bestPathRecursion(array, m, n, i, j + 1);
    return MAX(down, right);
}

int bestPath(int *array, int m, int n) {
    return bestPathRecursion(array, m, n, 0, 0);     
}
        
typedef struct {
    int result, m, n;
    int *array;
} TestCase;

int arr1[] = {9};
int arr2[] = {1, 2, 3};
int arr3[] = {1, 2, 3,
              4, 5, 6,
              7, 8, 9};
int arr4[] = {2, 2, 1, 1, 1,
              1, 2, 1, 1, 1,
              1, 2, 2, 1, 1, 
              1, 1, 2, 2, 1,
              1, 1, 1, 2, 2};  
TestCase tests[] = {
    {9, 1, 1, arr1}, 
    {6, 1, 3, arr2}, 
    {6, 3, 1, arr2}, 
    {29, 3, 3, arr3}, 
    {18, 5, 5, arr4} 
};  

#define NUMBER_OF_TESTS sizeof(tests) / sizeof(TestCase)

void test() {
    int result;
    int errors = 0;
    for(int i = 0; i < NUMBER_OF_TESTS; ++i) {
        result = bestPath(tests[i].array, tests[i].m, tests[i].n);
        printf("%d\n", result);
        if (result != tests[i].result) {
            printf("Error: Best path of test case number %d is %d. Got %d instead\n",
                i, tests[i].result, result);
            errors++;
        } 
    }

    if (errors)
        printf("Got %d errors\n", errors);
    else
        printf("Good work\n");
}

int main() {
    test();
    return 0;
} 

File: Recursion/BestPath.java
/******************************************************************************
 *
 *  Find best path in an MxN matrix
 *
 *****************************************************************************/
public class BestPath {
    public static void main(String [] args) {
        int result;
        int errors = 0;
        for(int i = 0; i < tests.length; ++i) {
            result = bestPath(tests[i].array);
            System.out.println(result);
            if (result != tests[i].result) {
                System.out.println(
                        "Error: Best path of test case number " + i + " is " + tests[i].result + ". Got " + result + " instead");
                errors++;
            }
        }

        if (errors > 0)
            System.out.println("Got " + errors + " errors");
        else
            System.out.println("Good work");

    }

    public static int bestPath(int[][] array) {
        return bestPath(array, 0, 0);
    }

    private static int bestPath(int[][] array, int i, int j) {
        int m = array.length;
        int n = array[0].length;
        if (i == m - 1 && j == n - 1)
            return array[i][j];

        if (i == m - 1)
            return array[i][j] + bestPath(array, i, j + 1);

        if (j == n - 1)
            return array[i][j] + bestPath(array, i + 1, j);

        int down = array[i][j] + bestPath(array, i + 1, j);
        int right = array[i][j] + bestPath(array, i, j + 1);
        return Math.max(down, right);
    }

    private static class TestCase {
        public TestCase(int[][] a, int r) {
            array = a;
            result = r;
        }
        int result;
        int[][] array;
    }

    private static int[][] arr1 = {{9}};
    private static int[][] arr2 = {{1, 2, 3}};
    private static int[][] arr3 = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    private static int[][] arr4 = {{2, 2, 1, 1, 1}, {1, 2, 1, 1, 1}, {1, 2, 2, 1, 1}, {1, 1, 2, 2, 1}, {1, 1, 1, 2, 2}};
    private static TestCase tests[] = {
            new TestCase(arr1, 9),
            new TestCase(arr2, 6),
            new TestCase(arr3, 29),
            new TestCase(arr4, 18)
    };
}
